gcsio/src/main/java/com/google/cloud/hadoop/gcsio/BatchExecutor.java

```diff
@@ -0,0 +1,147 @@
+            new ThreadFactoryBuilder()
+                .setNameFormat("gcs-manual-batching-pool-%d")
+                .setDaemon(true)
+                .build());
+    executor.allowCoreThreadTimeOut(true);
+    executor.setRejectedExecutionHandler(new ThreadPoolExecutor.CallerRunsPolicy());
+    return executor;
+  }
+
+  /** Adds a task to the execution queue. */
+  public <T> void queue(Callable<T> task, JsonBatchCallback<T> callback) {
+    checkState(
+        !requestsExecutor.isShutdown() && !requestsExecutor.isTerminated(),
+        "requestExecutor should not be terminated to queue request");
+
+    responseFutures.add(
+        requestsExecutor.submit(
+            () -> {
+              execute(task, callback);
+              return null;
+            }));
+  }
+
+  private static <T> void execute(Callable<T> task, JsonBatchCallback<T> callback)
+      throws Exception {
+    try {
+      T res = task.call();
+      callback.onSuccess(res, /* httpHeaders */ null);
+    } catch (IOException e) {
+      GoogleJsonResponseException jsonException = ApiErrorExtractor.getJsonResponseException(e);
+      if (jsonException == null) {
+        throw e;
+      }
+      callback.onFailure(jsonException.getDetails(), jsonException.getHeaders());
+    }
+  }
+
+  /** Awaits until all tasks are terminated and then shutdowns the executor. */
+  public void shutdown() throws IOException {
+    try {
+      awaitRequestsCompletion();
+      checkState(responseFutures.isEmpty(), "responseFutures should be empty after flush");
+    } finally {
+      requestsExecutor.shutdown();
+      try {
+        if (!requestsExecutor.awaitTermination(1, TimeUnit.SECONDS)) {
+          logger.atWarning().log("Forcibly shutting down manual batching thread pool.");
+          requestsExecutor.shutdownNow();
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        logger.atFine().withCause(e).log(
+            "Failed to await termination: forcibly shutting down manual batching thread pool.");
+        requestsExecutor.shutdownNow();
+      }
+    }
+  }
+
+  /** Awaits until all sent requests are completed. Should be serialized */
+  private void awaitRequestsCompletion() throws IOException {
+    while (!responseFutures.isEmpty()) {
+      getFromFuture(responseFutures.remove());
+    }
+  }
+
+  <T> T getFromFuture(Future<T> future) throws IOException {
+    try {
+      return future.get();
+    } catch (ExecutionException | InterruptedException e) {
+      GoogleCloudStorageEventBus.postOnException();
+      if (e instanceof InterruptedException) {
+        Thread.currentThread().interrupt();
+      }
+      logger.atSevere().withCause(e).log("Forcibly shutting down manual batching thread pool.");
+      requestsExecutor.shutdownNow();
+      throw new IOException(
+          String.format(
+              "Failed to get result: %s", e instanceof ExecutionException ? e.getCause() : e),
```

Comment:
Failed to delete folder. I assume the error message has details about the folder which failed.

Reply:
I was not under impression to make these errors specific for folders and in future this class might be used for bunch of different task, WDYT ?